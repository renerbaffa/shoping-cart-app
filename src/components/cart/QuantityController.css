.QuantityController-container {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.QuantityController-buttons-container {
  display: flex;
  align-items: center;
  margin-top: 4px;
}

.QuantityController-less-button {
  padding: 0px 8px;
  margin-right: 6px;
}

.QuantityController-more-button {
  padding: 0px 6px;
  margin-left: 6px;
}
