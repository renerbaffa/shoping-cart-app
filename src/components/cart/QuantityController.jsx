import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import Button from '../shared/Button';

import './QuantityController.css';

const QuantityController = ({
  className,
  max,
  min,
  onAddItem,
  onRemoveItem,
  quantity,
  ...other,
}) => (
  <div
    {...other}
    className={cx('QuantityController-container', className)}
  >
    <div><b>Quantity</b></div>
    <div
      className="QuantityController-buttons-container"
    >
      <Button
        className="QuantityController-less-button"
        disabled={quantity <= min}
        onClick={onRemoveItem}
      >
        -
      </Button>
      <div>{quantity}</div>
      <Button
        className="QuantityController-more-button"
        onClick={onAddItem}
        disabled={quantity >= max}
      >
        +
      </Button>
    </div>
  </div>
);

QuantityController.propTypes = {
  className: PropTypes.string,
  max: PropTypes.number,
  min: PropTypes.number,
  onAddItem: PropTypes.func,
  onRemoveItem: PropTypes.func,
  quantity: PropTypes.number,
};

QuantityController.defaultProps = {
  className: '',
  max: 0,
  min: 0,
  onAddItem: () => {},
  onRemoveItem: () => {},
  quantity: 0,
};

export default QuantityController;
