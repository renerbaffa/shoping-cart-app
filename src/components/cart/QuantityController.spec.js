import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import QuantityController from './QuantityController';

describe('<QuantityController />', () => {
  let wrapper;
  let onRemoveItem;
  let onAddItem;

  beforeEach(() => {
    onRemoveItem = jest.fn();
    onAddItem = jest.fn();
    wrapper = shallow(<QuantityController quantity={12} />);
  });

  it('should render correctly', () => {
    const quantityCOntroller = renderer.create(
      <QuantityController
        min={0}
        max={20}
        onAddItem={onAddItem}
        onRemoveItem={onRemoveItem}
        quantity={12}
      />
    ).toJSON();
    expect(quantityCOntroller).toMatchSnapshot();
  });

  describe('when min quantity is provided', () => {
    beforeEach(() => {
      wrapper.setProps({ min: 0 });
    });

    describe('when quantity is grater than min', () => {
      it('should not set disable to less button', () => {
        expect(
          wrapper.find('.QuantityController-less-button').props().disabled
        ).toBeFalsy();
      });

      it('should call onRemoveItem from props', () => {
        wrapper.setProps({ onRemoveItem });
        expect(onRemoveItem).not.toHaveBeenCalled();
        wrapper.find('.QuantityController-less-button').simulate('click');
        expect(onRemoveItem).toHaveBeenCalled();
      });
    });

    describe('when quantity is equal to the min', () => {
      it('should display less button disabled', () => {
        wrapper.setProps({ min: 0, quantity: 0 });
        expect(
          wrapper.find('.QuantityController-less-button').props().disabled
        ).toBeTruthy();
      });
    });
  });

  describe('when max quantity is provided', () => {
    beforeEach(() => {
      wrapper.setProps({ max: 13 });
    });

    describe('when quantity is less than max', () => {
      it('should not set disable to more button', () => {
        expect(
          wrapper.find('.QuantityController-more-button').props().disabled
        ).toBeFalsy();
      });

      it('should call onAddItem from props', () => {
        wrapper.setProps({ onAddItem });
        expect(onAddItem).not.toHaveBeenCalled();
        wrapper.find('.QuantityController-more-button').simulate('click');
        expect(onAddItem).toHaveBeenCalled();
      });
    });

    describe('when quantity is equal to max', () => {
      it('should display more button disabled', () => {
        wrapper.setProps({ max: 12 });
        expect(
          wrapper.find('.QuantityController-more-button').props().disabled
        ).toBeTruthy();
      });
    });
  });

  it('should reach 100% of coverage', () => {
    QuantityController.defaultProps.onRemoveItem();
    QuantityController.defaultProps.onAddItem();
  });
});
