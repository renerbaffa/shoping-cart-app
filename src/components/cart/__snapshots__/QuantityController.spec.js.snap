// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QuantityController /> should render correctly 1`] = `
<div
  className="QuantityController-container"
>
  <div>
    <b>
      Quantity
    </b>
  </div>
  <div
    className="QuantityController-buttons-container"
  >
    <button
      className="QuantityController-less-button Button"
      disabled={false}
      onClick={[Function]}
    >
      -
    </button>
    <div>
      12
    </div>
    <button
      className="QuantityController-more-button Button"
      disabled={false}
      onClick={[Function]}
    >
      +
    </button>
  </div>
</div>
`;
